# Section 1: Import Libraries and Load Data
This section imports necessary Python libraries and loads the traffic incidents dataset into a DataFrame for initial inspection.


In [2]:
# Import necessary libraries
import pandas as pd
import numpy as np

# Load the dataset
file_path = '../data/Traffic_Incidents_20241225.csv'
data = pd.read_csv(file_path)

# Display dataset information
print("Dataset Info:")
data.info()

# Display first 5 rows
print("\nFirst 5 Rows:")
data.head()

Dataset Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51756 entries, 0 to 51755
Data columns (total 10 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   INCIDENT INFO  51756 non-null  object 
 1   DESCRIPTION    51754 non-null  object 
 2   START_DT       51756 non-null  object 
 3   MODIFIED_DT    37699 non-null  object 
 4   QUADRANT       37697 non-null  object 
 5   Longitude      51756 non-null  float64
 6   Latitude       51756 non-null  float64
 7   Count          51756 non-null  int64  
 8   id             51756 non-null  object 
 9   Point          51756 non-null  object 
dtypes: float64(2), int64(1), object(7)
memory usage: 3.9+ MB

First 5 Rows:


INCIDENT INFO  \
0   Westbound 16 Avenue at Deerfoot Trail NE    
1                  11 Avenue and 4 Street SW    
2             68 Street and Memorial Drive E    
3       Eastbound 16 Avenue and 36 Street NE    
4              Barlow Trail and 61 Avenue SE    

                                         DESCRIPTION          START_DT  \
0  Stalled vehicle.  Partially blocking the right...   2022-06-21 7:31   
1          Traffic incident. Blocking multiple lanes   2022-06-21 4:02   
2                                  Traffic incident.  2022-06-20 23:53   
3       Traffic incident. Blocking the left shoulder  2022-06-20 16:43   
4                                  Traffic incident.  2022-06-20 16:42   

        MODIFIED_DT QUADRANT   Longitude   Latitude  Count  \
0   2022-06-21 7:33       NE -114.026687  51.067485      1   
1   2022-06-21 4:12       SW -114.071481  51.042624      1   
2  2022-06-20 23:55       NE -113.935553  51.052474      1   
3  2022-06-20 17:17       NE -113.989219  51.067086      1   
4  2022-06-20 17:28       SE -113.985727  50.998727      1   

                                                  id  \
0  2022-06-21T07:31:4051.067485129276236-114.0266...   
1  2022-06-21T04:02:1151.04262449261462-114.07148...   
2  2022-06-20T23:53:0851.0524735056658-113.935553...   
3  2022-06-20T16:43:2151.06708565896752-113.98921...   
4  2022-06-20T16:42:1250.99872748477766-113.98572...   

                                            Point  
0  POINT (-114.02668672232672 51.067485129276236)  
1   POINT (-114.07148057660925 51.04262449261462)  
2      POINT (-113.935553325751 51.0524735056658)  
3   POINT (-113.98921905311566 51.06708565896752)  
4   POINT (-113.98572655353505 50.99872748477766)

# Section 2: Check Missing Values
This section examines missing values in the dataset and plans how to handle them appropriately.

In [3]:
# Check for missing values
missing_values = data.isnull().sum()
print("Missing Values:\n", missing_values)

# Percentage of missing values
missing_percent = (missing_values / len(data)) * 100
print("\nPercentage of Missing Values:\n", missing_percent)

Missing Values:
 INCIDENT INFO        0
DESCRIPTION          2
START_DT             0
MODIFIED_DT      14057
QUADRANT         14059
Longitude            0
Latitude             0
Count                0
id                   0
Point                0
dtype: int64

Percentage of Missing Values:
 INCIDENT INFO     0.000000
DESCRIPTION       0.003864
START_DT          0.000000
MODIFIED_DT      27.160136
QUADRANT         27.164000
Longitude         0.000000
Latitude          0.000000
Count             0.000000
id                0.000000
Point             0.000000
dtype: float64


# Section 3: Handle Missing Values
In this section, we address missing values based on their nature:
1. 'DESCRIPTION' has 2 missing values, so we fill them with 'Unknown'.
2. 'MODIFIED_DT' has 27.16% missing values, which can be filled with 'START_DT' since it's the closest timestamp available.
3. 'QUADRANT' has 27.16% missing values, which can be filled with 'Unknown' as a placeholder for undefined locations.

In [7]:
# Handle missing values
data['DESCRIPTION'] = data['DESCRIPTION'].fillna('Unknown')  # Replace missing descriptions
data['MODIFIED_DT']= data['MODIFIED_DT'].fillna(data['START_DT'])  # Replace missing modified date with start date
data['QUADRANT'] = data['QUADRANT'].fillna('Unknown')  # Replace missing quadrant with 'Unknown'

# Verify no missing values remain
print("Missing Values After Handling:\n", data.isnull().sum())


Missing Values After Handling:
 INCIDENT INFO    0
DESCRIPTION      0
START_DT         0
MODIFIED_DT      0
QUADRANT         0
Longitude        0
Latitude         0
Count            0
id               0
Point            0
dtype: int64


# Section 4: Convert Date Columns and Extract Features
The date columns are converted into datetime objects to enable time-based analysis.  
New features like `Start_Date`, `Start_Hour`, and `Start_Weekday` are added to analyze temporal patterns of traffic incidents.

In [8]:
# Convert to datetime format
data['START_DT'] = pd.to_datetime(data['START_DT'])
data['MODIFIED_DT'] = pd.to_datetime(data['MODIFIED_DT'])

# Extract additional time-based features
data['Start_Date'] = data['START_DT'].dt.date
data['Start_Hour'] = data['START_DT'].dt.hour
data['Start_Weekday'] = data['START_DT'].dt.weekday  # 0 = Monday, 6 = Sunday

# Display modified data
print(data[['START_DT', 'MODIFIED_DT', 'Start_Date', 'Start_Hour', 'Start_Weekday']].head())

             START_DT         MODIFIED_DT  Start_Date  Start_Hour  \
0 2022-06-21 07:31:00 2022-06-21 07:33:00  2022-06-21           7   
1 2022-06-21 04:02:00 2022-06-21 04:12:00  2022-06-21           4   
2 2022-06-20 23:53:00 2022-06-20 23:55:00  2022-06-20          23   
3 2022-06-20 16:43:00 2022-06-20 17:17:00  2022-06-20          16   
4 2022-06-20 16:42:00 2022-06-20 17:28:00  2022-06-20          16   

   Start_Weekday  
0              1  
1              1  
2              0  
3              0  
4              0  


# Section 5: Advanced Data Cleaning
This section performs additional checks and cleaning to ensure data quality and consistency:
1. Check for duplicate rows and drop them if necessary.
2. Validate geographic coordinates (latitude and longitude) to ensure all values are valid.
3. Standardize text data, such as descriptions, to avoid inconsistencies.
4. Verify logical consistency in time fields (e.g., modified date >= start date).


In [10]:
# Check for duplicate rows
duplicates = data.duplicated()
print("Number of Duplicate Rows:", duplicates.sum())

# # Drop duplicates if any
# data = data.drop_duplicates()
# print("Duplicates removed. Remaining rows:", len(data))

Number of Duplicate Rows: 0


# Section 6: Validate Coordinates
We verify whether latitude and longitude values are within valid ranges:
- Latitude: -90 to 90
- Longitude: -180 to 180

In [11]:
# Check valid latitude and longitude ranges
invalid_lat = data[(data['Latitude'] < -90) | (data['Latitude'] > 90)]
invalid_lon = data[(data['Longitude'] < -180) | (data['Longitude'] > 180)]

print(f"Invalid latitude values: {len(invalid_lat)}")
print(f"Invalid longitude values: {len(invalid_lon)}")

# Drop rows with invalid coordinates if any
data = data[(data['Latitude'] >= -90) & (data['Latitude'] <= 90)]
data = data[(data['Longitude'] >= -180) & (data['Longitude'] <= 180)]

Invalid latitude values: 0
Invalid longitude values: 0


# Section 7: Detect Outliers in Time Features
We inspect the start and modified timestamps for any anomalies, such as future or very old dates.

In [12]:
# Check for outliers in datetime columns
print(f"Earliest Start Date: {data['START_DT'].min()}")
print(f"Latest Start Date: {data['START_DT'].max()}")

print(f"Earliest Modified Date: {data['MODIFIED_DT'].min()}")
print(f"Latest Modified Date: {data['MODIFIED_DT'].max()}")

# Filter out unreasonable dates if needed
# Example: Only keep incidents from the last 10 years
import datetime
start_limit = datetime.datetime.now() - pd.DateOffset(years=10)
data = data[data['START_DT'] >= start_limit]

Earliest Start Date: 2016-12-06 10:00:00
Latest Start Date: 2024-12-25 21:54:00
Earliest Modified Date: 2016-12-06 10:01:00
Latest Modified Date: 2024-12-25 21:55:00


# Section 8: Standardize Text Columns
We clean and standardize text columns such as 'DESCRIPTION' and 'QUADRANT' to ensure consistency and remove extra whitespace.


In [14]:
# Standardize 'DESCRIPTION' and 'QUADRANT'
data['DESCRIPTION'] = data['DESCRIPTION'].str.strip().str.lower()
data['QUADRANT'] = data['QUADRANT'].str.strip().str.upper()

# Verify changes
print(data[['DESCRIPTION', 'QUADRANT']].head())

                                         DESCRIPTION QUADRANT
0  stalled vehicle.  partially blocking the right...       NE
1          traffic incident. blocking multiple lanes       SW
2                                  traffic incident.       NE
3       traffic incident. blocking the left shoulder       NE
4                                  traffic incident.       SE


# Section 9: Save Cleaned Data
The cleaned dataset is saved as a CSV file to preserve progress and make it easier to load in future analysis steps.

In [16]:
# Save the cleaned dataset
data.to_csv('../data/Cleaned_Traffic_Incidents.csv', index=False)

print("Cleaned dataset saved successfully!")

Cleaned dataset saved successfully!
